README.md
# rlang-mocks

A mock-up cli script set of [R: The R Project for Statistical Computing](https://www.r-project.org/) that filter text-object input from the pipeline(stdin) and return text-object.

- For use in UTF-8 Japanese environments on windows.
- For my personal work and hobby use.
- Note that the code is spaghetti (due to my technical inexperience).
- Insufficient tests and error handlings.

script list:

```powershell
# one-liner to create function list for PowerShell
(cat README.md | sls '^#### \[[^[]+\]').Matches.Value.Replace('#### ','') -join ", " | Set-Clipboard
```

- [rcalc.R], [rmatcalc.R], [sketch.R]


主に現実世界の不定形文字列に対してパターンマッチング処理を行うためのフィルタ群。基本的な入力として、UTF-8＋半角スペース区切り＋行指向のパイプライン経由文字列データ（テキストオブジェクト）を期待する。

`src`下のファイルは1ファイル1関数。基本的に他の関数には依存しないようにしているので、関数ファイル単体を移動して利用することもできる。（一部の関数は他の関数ファイルに依存しているものもある）

**充分なエラー処理をしていない**モックアップ。事務職（非技術職）な筆者の毎日の仕事（おもに文字列処理）を、簡単便利に楽しくさばくための道具としてのコマンドセット。


## Install functions

1. Put `*.R` files under the `src` directory at any location.
2. Set terminal input/output encoding to `UTF-8`
    - The functions expect `UTF-8` encoded input, so if you want to run them on PowerShell in a Japanese environment, make sure the encoding is ready in advance.
    - if you use PowerShell, run the following dot sourcing command
        - `. path/to/rlang-mocks/operator.ps1`

関数群はUTF-8エンコードされた入力を期待するので、
関数実行前にカレントプロセスのエンコードを`UTF-8`にしておくとよい。

```powershell
# for PowerShell
# install favorite functions for japanese environment
# set encode
if ($IsWindows){
    chcp 65001
    [System.Console]::OutputEncoding = [System.Text.Encoding]::GetEncoding("utf-8")
    [System.Console]::InputEncoding  = [System.Text.Encoding]::GetEncoding("utf-8")
    # compartible with multi byte code
    $env:LESSCHARSET = "utf-8"
}
```

```powershell
# for PowerShell
# or sourcing dot files
. path/to/rlang-mocks/operator.ps1
```

## Description of each functions

各関数の挙動と作った動機と簡単な説明。

### Show functions

None

### Multipurpose

#### [rcalc.R] - Cli rscript executer

[rcalc.R]: src/rcalc.R

- Usage
    - man: `Rscript rcalc.R [-h]`
    - `Rscript rcalc.R -f <formula> [-d <delim>] [opts...]`
- Library
    - require: `optparse`
    - optional: `ggplot2`, `tidyverse`

Synopsis:

```powershell
$ Rscript rcalc.R
Usage: Rscript rcalc.R -f <formula> [-d <delim>] [opts...]
  --delim,      -d :delimiter
  --formula,    -f :formula
  --input,      -i :input data file
  --noheader       :no heder input
  --colnames,   -c :data column names separated with comma
  --colclasses     :data column clasees (character,integer,numeric,factor,etc...)
  --library,    -l :import libraries separated with comma
  --plot,       -p :show plot
  --output,     -o :output plot image
  --factor         :stringsAsFactors = TRUE
  --debug          :print debug
```

Examples:

```powershell
# simple usage
cat iris.csv | Rscript rcalc.R -f 'summary(df)' -d ','
```

```powershell
# rename column names
cat iris.csv | Rscript rcalc.R -f 'df |> head()' -d ',' -l 'tidyverse' -c sl,sw,pl,pw,sp
```

```powershell
# set colnames and colclasses
Rscript rcalc.R -f 'sapply(df, class)' -i iris.csv -d ',' -c 'sl,sw,pl,pw,sp' --colclasses 'numeric,numeric,numeric,numeric,factor'
         sl          sw          pl          pw          sp
  "numeric"   "numeric"   "numeric"   "numeric" "character"
```

```powershell
# use built-in examples
echo 1 | Rscript rcalc.R -f 'letters'
```

```powershell
# import libraries
cat iris.csv | Rscript rcalc.R -f 'summary(df)' -d ',' -l ggplot2,optparse
echo 1 | Rscript rcalc.R -f 'iris %>% group_by(Species) %>% summarise(mean = mean(Petal.Length))' -d ',' -l dplyr

# install libraries
Rscript -e 'install.packages("palmerpenguins", repos="https://cran.r-project.org/")'
echo 1 | Rscript rcalc.R -f 'install.packages("palmerpenguins", repos="https://cran.r-project.org/")'
```

```powershell
# output package startup message (-m|--message)
echo 1 | Rscript rcalc.R -f 'letters' -l tidyverse -m
```

```powershell
# output csv
cat iris.csv | Rscript rcalc.R -f 'summary(df);write.csv(df,"",quote=FALSE)' -d ','
```

```powershell
# plot
cat iris.csv | Rscript rcalc.R -f 'plot(df)' -d ',' --plot
```

```powershell
# ggplot2 : using print() and --plot
cat iris.csv | Rscript rcalc.R -d ',' -f "p <- ggplot(data=df,mapping=aes(x=sepal_length,y=sepal_width))+layer(geom='point',stat='identity',position='identity');print(p)" -l ggplot2 --plot
```

the above is equivalent to the following Rscript (RScript a.R).
note that unavailable multibyte character in rscript

```r
library(ggplot2)
X11()
p <- ggplot(data=iris,mapping=aes(x=Sepal.Length,y=Sepal.Width))+
    layer(geom='point',stat='identity',position='identity')
print(p)
Sys.sleep(1000)
dev.off()
```

```powershell
## ggplot2 another example: histogram
cat iris.csv | Rscript rcalc.R -d ',' -f "p <- ggplot(iris)+geom_histogram(aes(Petal.Length, fill=Species), binwidth=0.5)+facet_wrap(~Species);print(p)" -l ggplot2 --plot

## ggplot2 another example: histogram using gghighlight
cat iris.csv | Rscript rcalc.R -d ',' -f "p <- ggplot(iris)+geom_histogram(aes(Petal.Length, fill = Species), binwidth = 0.5)+gghighlight()+facet_wrap(~ Species);print(p)" -l ggplot2,gghighlight --plot

## ggplot2 another example: geom_point using gghighlight
cat iris.csv | Rscript rcalc.R -d ',' -f "p <- ggplot(data=df,mapping=aes(x=sepal_length,y=sepal_width,colour=species))+geom_point()+gghighlight(grepl('^v',species));print(p)" -l ggplot2,gghighlight --plot
```

```powershell
# save plot as image file
cat iris.csv | Rscript rcalc.R -f 'hist(df$sepal_length)' -d ',' -o a.png
cat iris.csv | Rscript rcalc.R -f 'hist(df$sepal_length)' -d ',' -o a.png -s 400,300
cat iris.csv | Rscript rcalc.R -f 'hist(df$sepal_length)' -d ',' -o a.jpg -s 400,300
cat iris.csv | Rscript rcalc.R -f 'hist(df$sepal_length)' -d ',' -o a.bmp -s 6,4
cat iris.csv | Rscript rcalc.R -f 'hist(df$sepal_length)' -d ',' -o a.pdf
cat iris.csv | Rscript rcalc.R -f 'hist(df$sepal_length)' -d ',' -o a.eps
```

```powershell
# eval external R source file

# install.packages("palmerpenguins")
# from: https://allisonhorst.github.io/palmerpenguins/
echo 1 | Rscript rcalc.R -f 'install.packages("palmerpenguins", repos="https://cran.r-project.org/")'

# eval palmerpenguins
echo 1 | Rscript rcalc.R -f a.R -l 'palmerpenguins,ggplot2' --plot
         Rscript rcalc.R -f a.R -l 'palmerpenguins,ggplot2' --plot -i penguins.csv -d ','
```

```powershell
## palmer penguins by allison horst
echo 1 | Rscript rcalc.R -f 'penguins %>% count(species)' -l 'palmerpenguins,tidyverse'
## # A tibble: 3 x 2
##   species       n
##   <fct>     <int>
## 1 Adelie      152
## 2 Chinstrap    68
## 3 Gentoo      124
```

```powershell
echo 1 | Rscript rcalc.R -f 'penguins %>% group_by(species) %>% summarize(across(where(is.numeric), mean, na.rm = TRUE))' -l 'palmerpenguins,tidyverse'
## # A tibble: 3 x 6
##   species   bill_length_mm bill_depth_mm flipper_length_mm body_mass_g  year
##   <fct>              <dbl>         <dbl>             <dbl>       <dbl> <dbl>
## 1 Adelie              38.8          18.3              190.       3701. 2008.
## 2 Chinstrap           48.8          18.4              196.       3733. 2008.
## 3 Gentoo              47.5          15.0              217.       5076. 2008.
```

```powershell
echo 1 | Rscript rcalc.R -f "flipper_hist <- ggplot(data=penguins, aes(x=flipper_length_mm))+geom_histogram(aes(fill=species), alpha = 0.5, position='identity')+scale_fill_manual(values=c('darkorange','purple','cyan4'))+theme_minimal()+labs(x='Flipper length (mm)', y='Frequency', title='Penguin flipper lengths');print(flipper_hist)" -l 'palmerpenguins,ggplot2' --plot
 `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.
Warning message:
Removed 2 rows containing non-finite values (stat_bin).
```

### Math

#### [rmatcalc.R] - Cli matrix calculator by connecting with pipes

[rmatcalc.R]: src/rmatcalc.R

- Usage
    - man: `Rscript rmatcalc.R [-h]`
    - `Rscript rmatcalc.R -f <formula> [-d <delim>] [opts...]`
- Inspired by
    - [Ryuichi Ueda and CIT Autonomous Robot Lab](https://b.ueda.tech/?post=00674)
        - [GitHub - ryuichiueda/PMAT: Pipe Oriented Matrix Calculator](https://github.com/ryuichiueda/PMAT)
    - Command: `matcalc`, `pmat`
- Dependency
    - require: `optparse`

Synopsis:

```powershell
$ Rscript rmatcalc.R
Usage: Rscript rmatcalc.R -f <formula> [-d <delim>] [opts...]
  --formula,    -f :formula
  --delim,      -d :delimiter
  --input,      -i :input data from file
  --dtype          :array data type. defalut=numeric
  --nowrap         :no wrap
  --debug          :print debug
```

```
# frequently usage
  A+B-A%*%B : sum, difference, product
  A * B     : Multiply element by element
  A %o% B   : cross product
  A %x% B   : kronecker product

# Functions
  t(A)
  solve(A)
  eigen(A)
  det(A)
  diag(n)
  diag(a1:an)
  sum(x^n)
  crossprod(A)
  crossprod(x,y)
  rowSums(A)
  solSums(A)
  rowMeans(A)
  colMeans(A)
  x[upper.tri(A)] <- n
  x[lower.tri(A)] <- n
```

Examples:

```powershell
# input example:
$ cat matrix
A 1 2
A 3 4
B 4 3
B 2 1

# calc example:
$ cat matrix | Rscript rmatcalc.R -f 'A%*%B'
A 1 2
A 3 4
B 4 3
B 2 1
C 8 5
C 20 13

$ cat matrix | Rscript rmatcalc.R -f 'A*B'
A 1 2
A 3 4
B 4 3
B 2 1
C 4 6
C 6 4
```

```powershell
# transpose:
$ cat matrix | Rscript rmatcalc.R -f 't(A)'
A 1 2
A 3 4
B 1 3
B 2 4

# solve:
$ cat matrix | Rscript rmatcalc.R -f 'A%*%solve(A)'
A 1 1
A 2 4
B 1 0
B 0 1

# add new label to ans:
$ cat matrix | Rscript rmatcalc.R -f 'C=A*B'
A 1 2
A 3 4
B 4 3
B 2 1
C 4 6
C 6 4
```

```powershell
# determinant:
#   To output a single (non-matrix) value,
#   multiply by diag(1)
$ cat matrix | Rscript rmatcalc.R -f 'C=det(A)*diag(1)'
A 2 -6 4
A 7 2 3
A 8 5 -1
C -144

# rank:
$ cat matrix | Rscript rmatcalc.R -f 'C=qr(A)$rank*diag(1)'
A 2 -6 4
A 7 2 3
A 8 5 -1
C 3

# chain calc using pipe:
$ cat matrix | Rscript rmatcalc.R -f 'C=A*B' | Rscript rmatcalc.R -f 'E=C%*%A'
A 1 2
A 3 4
B 4 3
B 2 1
C 4 6
C 6 4
E 22 32
E 18 28
```

### Image processing

#### [sketch.R] - A wrapper script of "sketcher" library

[sketch.R]: src/sketch.R

- Usage
    - man: `Rscript sketch.R [-h]`
    - `Rscript sketch.R -i a.png [--debug] [[param] [param] ... [param]]`
        - `--input, -i  :An input image`
        - `--style      :Sketch style: 1 or 2, defalut: 1`
        - `--lineweight :Strength of lines: >=0.3, default: 1`
        - `--smooth     :Smoothness of texture: >=0, default: 1`
        - `--gain,      :Gain parameter: betw 0 and 1, default: 0.02`
        - `--contrast   :Contrast parameter: >=0, default: 20(for style1) or 4(for style2)`
        - `--shadow     :Shadow threshold: betw 0 and 1, defalut 0.0`
        - `--maxsize    :Max resolution of output: >0, default: 2048`
        - `--output, -o :output file, default: NA`
        - `--debug      :print debug, default: FALSE`
- Dependency
    - require: `optparse`, `sketcher`

Examples:

```powershell
# Case1: Outline is missing and texture is lacking
Rscript sketch.R -i a.png --style 2 --shadow 0.4

# Case2: Due to the lack of edges in the dark region of the face
Rscript sketch.R -i a.png --shadow 0.4

# Case3: Neko. objects have unclear edges/outlines
Rscript sketch.R -i a.png --smooth 0
```

## CREDITS

### [rmatcalc.R]

- used functions:
    - `getIndex`, `append_VecOrList`, `extend_VecOrList`, `remove_vecOrList`, `insert_VecOrList`, `pop_VecOrList`, `getCount`
- from:
    - GitHub - usagi-san-dayo/UsagiSan/R/UsagiSan.R
        - <https://github.com/usagi-san-dayo/UsagiSan>
    - blog - usagi-san no toukeigaku salon
        - <https://multivariate-statistics.com/2021/01/29/r-programming-vector-list/>
- license:
    - MIT License: Copyright (c) 2020 UsagiSan

### [sketch.R]

- used library:
    - `sketcher`
- from:
    - GitHub - <https://github.com/tsuda16k/sketcher>
    - Blog - <https://htsuda.net/sketcher/>
- license:
    - MIT License: Copyright (c) 2020 Hiroyuki Tsuda



CHANGELOG.md
# Changelog

All notable changes to "rlang-mocks" project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Changed

- None

### Added

- None


## [1.0.2]

### Changed

- Updated [README.md]

### Added

- Added [rcalc.R] examples


## [1.0.1]


### Added

- Added Pester tests for [rmatcalc.R]


## [1.0.0] - 2023-02-28

### Added

- Added [CHANGELOG.md]

## [0.1.0] - 2023-02-28

### Changed

- Changed [README.md] section links

### Added 

- Added [rcalc.R], [rmatcalc.R], [sketch.R]


[README.md]: blob/main/README.md
[CHANGELOG.md]: blob/main/CHANGELOG.md

[rcalc.R]: src/rcalc.R
[rmatcalc.R]: src/rmatcalc.R
[sketch.R]: src/sketch.R



[unreleased]: https://github.com/btklab/rlang-mocks/compare/1.0.2..HEAD
[1.0.2]: https://github.com/btklab/rlang-mocks/releases/tag/1.0.2
[1.0.1]: https://github.com/btklab/rlang-mocks/releases/tag/1.0.1
[1.0.0]: https://github.com/btklab/rlang-mocks/releases/tag/1.0.0
[0.1.0]: https://github.com/btklab/rlang-mocks/releases/tag/0.1.0



src/rcalc.R
#!/usr/bin/env Rscript

# args
#args <- commandArgs(trailingOnly=TRUE)

# load file
#install.packages("optparse", dependencies = T)
#install.packages("Rtools", dependencies = T)
suppressWarnings(library('optparse'))

# check opt
argv = commandArgs(TRUE)
if ( is.na(argv[1])){
    cat("rcalc.R -- exec/eval rscript from cli", "\n", sep="")
    cat("", "\n", sep="")
    cat("Synopsis:", "\n", sep="")
    cat("", "\n", sep="")
    cat("Usage: Rscript rcalc.R -f <formula> [-d <delim>] [opts...]", "\n", sep="")
    cat("  --delim,      -d :delimiter", "\n", sep="")
    cat("  --formula,    -f :formula", "\n", sep="")
    cat("  --input,      -i :input data file", "\n", sep="")
    cat("  --noheader       :no heder input", "\n", sep="")
    cat("  --colnames,   -c :data column names separated with comma", "\n", sep="")
    cat("  --colclasses     :data column clasees (character,integer,numeric,factor,etc...)", "\n", sep="")
    cat("  --library,    -l :import libraries separated with comma", "\n", sep="")
    cat("  --plot,       -p :show plot", "\n", sep="")
    cat("  --output,     -o :output plot image", "\n", sep="")
    cat("  --factor         :stringsAsFactors = TRUE", "\n", sep="")
    cat("  --debug          :print debug", "\n", sep="")
    cat("", "\n", sep="")
    cat("Examples:", "\n", sep="")
    cat("# simple usage", "\n", sep="")
    cat("cat iris.csv | Rscript rcalc.R -f 'summary(df)' -d ','", "\n", sep="")
    cat("cat iris.csv | Rscript rcalc.R -f 'write.table(summary(df),sep=\",\")' -d ','", "\n", sep="")
    cat("", "\n", sep="")
    cat("# rename column names", "\n", sep="")
    cat("cat iris.csv | Rscript rcalc.R -f 'df |> head()' -d ',' -l 'tidyverse' -c sl,sw,pl,pw,sp", "\n", sep="")
    cat("", "\n", sep="")
    cat("# set colnames and colclasses", "\n", sep="")
    cat("Rscript rcalc.R -f 'sapply(df, class)' -i iris.csv -d ',' -c 'sl,sw,pl,pw,sp' --colclasses 'numeric,numeric,numeric,numeric,factor'", "\n", sep="")
    cat('         sl          sw          pl          pw          sp',"\n", sep="")
    cat('  "numeric"   "numeric"   "numeric"   "numeric" "character"',"\n", sep="")
    cat("", "\n", sep="")
    cat("", "\n", sep="")
    cat("# use built-in examples", "\n", sep="")
    cat("echo 1 | Rscript rcalc.R -f 'letters'", "\n", sep="")
    cat("", "\n", sep="")
    cat("# import libraries", "\n", sep="")
    cat("cat iris.csv | Rscript rcalc.R -f 'summary(df)' -d ',' -l ggplot2,optparse", "\n", sep="")
    cat("echo 1 | Rscript rcalc.R -f 'iris %>% group_by(Species) %>% summarise(mean = mean(Petal.Length))' -d ',' -l dplyr", "\n", sep="")
    cat("", "\n", sep="")
    cat("# install libraries", "\n", sep="")
    cat("Rscript -e 'install.packages(\"palmerpenguins\", repos=\"https://cran.r-project.org/\")'", "\n", sep="")
    cat("echo 1 | Rscript rcalc.R -f 'install.packages(\"palmerpenguins\", repos=\"https://cran.r-project.org/\")'", "\n", sep="")
    cat("", "\n", sep="")
    cat("# output package startup message (-m | --message)", "\n", sep="")
    cat("echo 1 | Rscript rcalc.R -f 'letters' -l tidyverse -m", "\n", sep="")
    cat("", "\n", sep="")
    cat("# output csv", "\n", sep="")
    cat("cat iris.csv | Rscript rcalc.R -f 'summary(df);write.csv(df,\"\",quote=FALSE)' -d ','", "\n", sep="")
    cat("", "\n", sep="")
    cat("# plot", "\n", sep="")
    cat("cat iris.csv | Rscript rcalc.R -f 'plot(df)' -d ',' --plot", "\n", sep="")
    cat("", "\n", sep="")
    cat("# ggplot2 : using print() and --plot", "\n", sep="")
    cat("cat iris.csv | Rscript rcalc.R -d ',' -f \"p <- ggplot(data=df,mapping=aes(x=sepal_length,y=sepal_width))+layer(geom='point',stat='identity',position='identity');print(p)\" -l ggplot2 --plot", "\n", sep="")
    cat("", "\n", sep="")
    cat("    the above is equivalent to the following Rscript (RScript a.R)", "\n", sep="")
    cat("    note that unavailable multibyte character in rscript", "\n", sep="")
    cat("", "\n", sep="")
    cat("    library(ggplot2)", "\n", sep="")
    cat("    X11()", "\n", sep="")
    cat("    p <- ggplot(data=iris,mapping=aes(x=Sepal.Length,y=Sepal.Width))+", "\n", sep="")
    cat("        layer(geom='point',stat='identity',position='identity')", "\n", sep="")
    cat("    print(p)", "\n", sep="")
    cat("    Sys.sleep(1000)", "\n", sep="")
    cat("    dev.off()", "\n", sep="")
    cat("", "\n", sep="")
    cat("## ggplot2 another example: histogram", "\n", sep="")
    cat("echo 1 | Rscript rcalc.R -d ',' -f \"p <- ggplot(iris)+geom_histogram(aes(Petal.Length, fill=Species), binwidth=0.5)+facet_wrap(~Species);print(p)\" -l ggplot2 --plot", "\n", sep="")
    cat("", "\n", sep="")
    cat("## ggplot2 another example: histogram using gghighlight", "\n", sep="")
    cat("echo 1 | Rscript rcalc.R -d ',' -f \"p <- ggplot(iris)+geom_histogram(aes(Petal.Length, fill = Species), binwidth = 0.5)+gghighlight()+facet_wrap(~ Species);print(p)\" -l ggplot2,gghighlight --plot", "\n", sep="")
    cat("", "\n", sep="")
    cat("## ggplot2 another example: geom_point using gghighlight", "\n", sep="")
    cat("cat iris.csv | Rscript rcalc.R -d ',' -f \"p <- ggplot(data=df,mapping=aes(x=sepal_length,y=sepal_width,colour=species))+geom_point()+gghighlight(grepl('^v',species));print(p)\" -l ggplot2,gghighlight --plot", "\n", sep="")
    cat("", "\n", sep="")
    cat("## ggplot2 another example: ggpairs", "\n", sep="")
    cat("1 | Rscript rcalc.R -f 'print( iris %>% dplyr::select(Sepal.Length, Petal.Length, Species) %>% ggpairs(mapping = aes(color = Species), lower = list(continuous = wrap(\"smooth\", size = 1.5, alpha = 0.8)), upper = list(continuous = wrap(\"cor\", size = 5)), diag = list(continuous = wrap(\"densityDiag\", alpha = 0.5)))+ggtitle(\"hoge\") )' -l 'tidyverse,GGally' --plot", "\n", sep="")
    cat("", "\n", sep="")
    cat("", "\n", sep="")
    cat("# save plot as image file", "\n", sep="")
    cat("cat iris.csv | Rscript rcalc.R -f 'hist(df$sepal_length)' -d ',' -o a.png", "\n", sep="")
    cat("cat iris.csv | Rscript rcalc.R -f 'hist(df$sepal_length)' -d ',' -o a.png -s 400,300", "\n", sep="")
    cat("cat iris.csv | Rscript rcalc.R -f 'hist(df$sepal_length)' -d ',' -o a.jpg -s 400,300", "\n", sep="")
    cat("cat iris.csv | Rscript rcalc.R -f 'hist(df$sepal_length)' -d ',' -o a.bmp -s 6,4", "\n", sep="")
    cat("cat iris.csv | Rscript rcalc.R -f 'hist(df$sepal_length)' -d ',' -o a.pdf", "\n", sep="")
    cat("cat iris.csv | Rscript rcalc.R -f 'hist(df$sepal_length)' -d ',' -o a.eps", "\n", sep="")
    cat("", "\n", sep="")
    cat("", "\n", sep="")
    cat("# eval external R source file", "\n", sep="")
    cat("# install.packages('palmerpenguins', repos='https://cran.r-project.org/')", "\n", sep="")
    cat("# from: https://allisonhorst.github.io/palmerpenguins/", "\n", sep="")
    cat("echo 1 | Rscript rcalc.R -f 'install.packages(\"palmerpenguins\", repos=\"https://cran.r-project.org/\")'", "\n", sep="")
    cat("", "\n", sep="")
    cat("echo 1 | Rscript rcalc.R -f a.R -l 'palmerpenguins,ggplot2' --plot", "\n", sep="")
    cat("         Rscript rcalc.R -f a.R -l 'palmerpenguins,ggplot2' --plot -i penguins.csv -d ','", "\n", sep="")
    cat("", "\n", sep="")
    cat("## palmer penguins by allison horst", "\n", sep="")
    cat("echo 1 | Rscript rcalc.R -f 'penguins %>% count(species)' -l 'palmerpenguins,tidyverse'", "\n", sep="")
    cat("## # A tibble: 3 x 2", "\n", sep="")
    cat("##   species       n", "\n", sep="")
    cat("##   <fct>     <int>", "\n", sep="")
    cat("## 1 Adelie      152", "\n", sep="")
    cat("## 2 Chinstrap    68", "\n", sep="")
    cat("## 3 Gentoo      124", "\n", sep="")
    cat("", "\n", sep="")
    cat("", "\n", sep="")
    cat("echo 1 | Rscript rcalc.R -f 'penguins %>% group_by(species) %>% summarize(across(where(is.numeric), mean, na.rm = TRUE))' -l 'palmerpenguins,tidyverse'", "\n", sep="")
    cat("## # A tibble: 3 x 6", "\n", sep="")
    cat("##   species   bill_length_mm bill_depth_mm flipper_length_mm body_mass_g  year", "\n", sep="")
    cat("##   <fct>              <dbl>         <dbl>             <dbl>       <dbl> <dbl>", "\n", sep="")
    cat("## 1 Adelie              38.8          18.3              190.       3701. 2008.", "\n", sep="")
    cat("## 2 Chinstrap           48.8          18.4              196.       3733. 2008.", "\n", sep="")
    cat("## 3 Gentoo              47.5          15.0              217.       5076. 2008.", "\n", sep="")
    cat("", "\n", sep="")
    cat("", "\n", sep="")
    cat("echo 1 | Rscript rcalc.R -f \"flipper_hist <- ggplot(data=penguins, aes(x=flipper_length_mm))+geom_histogram(aes(fill=species), alpha = 0.5, position='identity')+scale_fill_manual(values=c('darkorange','purple','cyan4'))+theme_minimal()+labs(x='Flipper length (mm)', y='Frequency', title='Penguin flipper lengths');print(flipper_hist)\" -l 'palmerpenguins,ggplot2' --plot", "\n", sep="")
    cat(" `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.", "\n", sep="")
    cat("Warning message:", "\n", sep="")
    cat("Removed 2 rows containing non-finite values (stat_bin).", "\n", sep="")
    cat("", "\n", sep="")
    cat("", "\n", sep="")
    quit(status=1)
}

optslist <- list(
    #make_option("--flag1", action="store_true", default=FALSE, help="A Flag"),
    make_option(c("-f", "--formula"),  type="character",    default=NA,    help="formula"),
    make_option(c("-d", "--delim"),    type="character",    default=' ',   help="delimiter"),
    make_option(c("-i", "--input"),    type="character",    default=NA,    help="An input image"),
    make_option(c("-c", "--colnames"), type="character",    default=NA,    help="column names"),
    make_option("--colclasses"       , type="character",    default=NA,    help="column classes"),
    make_option("--factor"           , action="store_true", default=FALSE, help="stringsAsFactors"),
    make_option(c("-o", "--output"),   type="character",    default=NA,    help="output file"),
    make_option(c("-l", "--library"),  type="character",    default=' ',   help="import libraries"),
    make_option(c("-p", "--plot"),     action="store_true", default=FALSE, help="no header input"),
    make_option(c("-s", "--size"),     type="character",    default=NA,    help="plot size"),
    make_option("--skip",              type="integer",      default=0,     help="skip rows"),
    make_option("--noheader",          action="store_true", default=FALSE, help="no header input"),
    make_option("--nowrap",            action="store_true", default=FALSE, help="no wrap output"),
    make_option("--na",                type="character",    default='NA',  help="na string"),
    make_option(c("-m", "--message"),  action="store_true", default=FALSE, help="output package startup message"),
    make_option("--debug",             action="store_true", default=FALSE, help="debug print")
    )
parser <- OptionParser(option_list=optslist)
opts <- parse_args(parser)
# test opts
## check input is provided
if (is.na(opts$formula)) {
    warning("formula(-f, --formula) parameter must be provided. See script usage (--help)")
    quit(status=1)
}

# check if output is provided
ofile = NA
if ( ! is.na(opts$output)){
    ## get extension of image file
    ofile  = opts$output
    ofile  = gsub("\\\\", "/", ofile)
    ofname = gsub("\\.[^.]+$", "", ofile)
    ofext  = tolower(gsub("^.*\\.", "", ofile))
    ## set image size
    if( ! is.na(opts$size)){
        s <- unlist(strsplit(opts$size, split=","))
        if(length(s) == 1){
            im_width  = strtoi(s[1])
            im_height = strtoi(s[1])
        }else{
            im_width  = strtoi(s[1])
            im_height = strtoi(s[2])
        }
    }else if (ofext == "png" || ofext == "jpg" || ofext == "jpeg") {
        im_width  = 600
        im_height = 600
    }else{
        im_width  = 6
        im_height = 6
    }

    im_comstr = NA
    if (ofext == "png") {
        im_comstr <- paste0(
        "png(file='", ofile, "'",
        ", width=", im_width,
        ", height=", im_height,
        ')')
    } else if (ofext == "jpg" || ofext == "jpeg") {
        im_comstr <- paste0(
        "jpeg(file='", ofile, "'",
        ", width=", im_width,
        ", height=", im_height,
        ')')
    } else if (ofext == "bmp") {
        im_width  = 7
        im_height = 7
        im_comstr <- paste0(
        "bmp(file='", ofile, "'",
        ", width=", im_width,
        ", height=", im_height,
        ')')
    } else if (ofext == "pdf") {
        im_comstr <- paste0(
        "pdf(file='", ofile, "'",
        #", width=", im_width,
        #", height=", im_height,
        ')')
    } else if (ofext == "eps") {
        im_width  = 7
        im_height = 7
        im_comstr <- paste0(
        "postscript(file='", ofile, "'",
        ', horizontal=F',
        ', onefile=F',
        ', paper="special"',
        ')')
    } else if (ofext == "txt" || ofext == "csv" || ofext == "tsv") {
        im_comstr <- NA
    } else {
        warning(paste(ofile, " is not available image format. jpg,png,bmp,pdf,eps is available.", sep = ""))
        quit(status=1)
    }
}

# debug print
if (opts$debug){
    if(!is.na(opts$formula)){cat("formula   :", opts$formula, "\n")}
    if(!is.na(opts$input))  {cat("input     :", opts$input, "\n")}
    if(!is.na(opts$output)) {cat("output    :", opts$output, "\t[output image file(.jpg or .png)]\n")}
    if(!is.na(opts$library)){cat("library   :", opts$library, "\n")}
    cat("debug     :", opts$debug,      "\t[bool, default: FALSE]\n")
}

# import libraries
import_libraries <- function (libstr){
    lopts <- ", warn.conflicts=FALSE"
    if (opts$message){
        pre <- ""
        suc <- ')'
    }else{
        pre <- "invisible(suppressPackageStartupMessages("
        suc <- ')))'
    }
    eval(parse(text=paste0(
        pre,
        "suppressWarnings(",
        "library(", libstr, lopts, ")",
        suc
        )))
}
if ( ! is.na(opts$library)){
    libs = unlist(strsplit(opts$library, split=","))
    import_libraries(libs)
}

# main
eval_formula <- function (evstr){
    ffile  = opts$formula
    ffile  = gsub("\\\\", "/", ffile)
    if (file.exists(ffile)) {
        ## if formula is R script file
        source(ffile)
    } else {
        eval(parse(text=(evstr)))
    }
}
if ( ! is.na(opts$input)){
    ifile = opts$input
    ifile = gsub("\\\\", "/", ifile)
    # test input
    if ( ! file.exists(ifile)) {
        warning(sprintf("Specified file ( %s ) does not exist", ifile))
        quit(status=1)
    }
    con = file(ifile, open="r")
}else{
    con = file("stdin", open="r")
}
if (opts$noheader){
    hflag = FALSE
}else{
    hflag = TRUE
}

# read table from file or stdin
if ( ! is.na(opts$colclasses)){
    colclasses = unlist(strsplit(opts$colclasses, split=","))
    df <- read.table(con, sep = opts$delim, header = hflag, quote = "",
        skip = opts$skip,
        na.strings = opts$na,
        colClasses = colclasses,
        stringsAsFactors = FALSE,
        fileEncoding = 'utf-8')
} else {
    df <- read.table(con, sep = opts$delim, header = hflag, quote = "",
        skip = opts$skip,
        na.strings = opts$na,
        stringsAsFactors = FALSE,
        fileEncoding = 'utf-8')
}
close(con)

# change column names
if ( ! is.na(opts$colnames)){
    colnames = unlist(strsplit(opts$colnames, split=","))
    colnames(df) <- colnames
}
if ((opts$plot) & (is.na(ofile))){
    # show plot
    X11()
    eval_formula(opts$formula)
    cat("Press Ctrl+c to continue...")
    Sys.sleep(1000)
    dev.off()
} else if (! is.na(ofile)){
    if (ofext == "txt" || ofext == "csv" || ofext == "tsv") {
        x <- eval_formula(opts$formula)
        write.table(x, ofile, sep=opts$delim, fileEncoding='UTF-8', quote=F, row.names=F, col.names=T, append=F)
    } else {
        # save plot as image file
        eval_formula(im_comstr)
        #print(im_comstr)
        eval_formula(opts$formula)
        dev.off()
        # check output file
        if (file.exists(ofile)) {
            print(paste("output:", ofile, sep = " "))
        } else {
            warning(paste("error: can't export to", ofile, sep = " "))
            quit(status=1)
        }
    }
} else if (opts$nowrap){
        x <- eval_formula(opts$formula)
        ofile = ""
        write.table(x, ofile, sep=opts$delim, fileEncoding='UTF-8', quote=F, row.names=F, col.names=T, append=F)
}else{
    eval_formula(opts$formula)
}

quit(status=0)
